File: examples/python/sdk_error_rate_limit_logs.py
"""Production-style ViralAPI SDK wrapper with bounded retry and logs.

Set VIRALAPI_API_KEY before running. The base URL remains OpenAI-compatible.
"""

from __future__ import annotations

import json
import os
import time
import uuid
from dataclasses import dataclass
from typing import Iterable

from openai import APIConnectionError, APIStatusError, APITimeoutError, OpenAI, RateLimitError


@dataclass(frozen=True)
class Route:
    model: str
    cost_group: str
    timeout: float


ROUTES: dict[str, list[Route]] = {
    "ai_support": [
        Route("claude-sonnet-4", "stable_official", 12),
        Route("gpt-4o-mini", "official_transfer", 10),
    ],
    "content_batch": [
        Route("gemini-2.5-flash", "welfare", 25),
        Route("gpt-4o-mini", "official_transfer", 20),
    ],
}

RETRYABLE_ERRORS = (APIConnectionError, APITimeoutError, RateLimitError)


def build_client() -> OpenAI:
    return OpenAI(
        api_key=os.environ["VIRALAPI_API_KEY"],
        base_url=os.environ.get("VIRALAPI_BASE_URL", "https://viralapi.ai/v1"),
        timeout=float(os.environ.get("VIRALAPI_TIMEOUT", "20")),
    )


def log_event(**fields: object) -> None:
    print(json.dumps(fields, ensure_ascii=False, sort_keys=True))


def complete_with_fallback(
    *,
    client: OpenAI,
    tenant_id: str,
    scenario: str,
    messages: Iterable[dict[str, str]],
) -> str:
    request_id = str(uuid.uuid4())
    routes = ROUTES[scenario]
    message_list = list(messages)

    for fallback_index, route in enumerate(routes):
        for attempt in range(1, 3):
            started = time.monotonic()
            try:
                response = client.chat.completions.create(
                    model=route.model,
                    messages=message_list,
                    timeout=route.timeout,
                )
                log_event(
                    event="llm_call_ok",
                    request_id=request_id,
                    tenant_id=tenant_id,
                    scenario=scenario,
                    model=route.model,
                    cost_group=route.cost_group,
                    fallback_index=fallback_index,
                    attempt=attempt,
                    latency_ms=round((time.monotonic() - started) * 1000),
                )
                return response.choices[0].message.content or ""
            except RETRYABLE_ERRORS as exc:
                log_event(
                    event="llm_call_retryable_error",
                    request_id=request_id,
                    tenant_id=tenant_id,
                    scenario=scenario,
                    model=route.model,
                    cost_group=route.cost_group,
                    fallback_index=fallback_index,
                    attempt=attempt,
                    error_type=exc.__class__.__name__,
                )
                time.sleep(min(2**attempt, 6))
            except APIStatusError as exc:
                log_event(
                    event="llm_call_status_error",
                    request_id=request_id,
                    tenant_id=tenant_id,
                    scenario=scenario,
                    model=route.model,
                    cost_group=route.cost_group,
                    fallback_index=fallback_index,
                    attempt=attempt,
                    status_code=exc.status_code,
                )
                if exc.status_code >= 500:
                    break
                raise

    raise RuntimeError(f"all LLM routes failed: request_id={request_id}")


if __name__ == "__main__":
    client = build_client()
    answer = complete_with_fallback(
        client=client,
        tenant_id="tenant_demo",
        scenario="content_batch",
        messages=[{"role": "user", "content": "Draft a concise support macro."}],
    )
    print(answer)

File: examples/node/sdk-error-rate-limit-logs.mjs
import crypto from "node:crypto";
import OpenAI from "openai";

const client = new OpenAI({
  apiKey: process.env.VIRALAPI_API_KEY,
  baseURL: process.env.VIRALAPI_BASE_URL || "https://viralapi.ai/v1",
  timeout: Number(process.env.VIRALAPI_TIMEOUT_MS || 20000),
});

const routes = {
  internal_tool: [
    { model: "gemini-2.5-flash", group: "welfare", timeoutMs: 25000 },
    { model: "gpt-4o-mini", group: "official_transfer", timeoutMs: 18000 },
  ],
  paid_saas_feature: [
    { model: "claude-sonnet-4", group: "stable_official", timeoutMs: 12000 },
    { model: "gpt-4o-mini", group: "official_transfer", timeoutMs: 10000 },
  ],
};

function sleep(ms) {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

function logEvent(fields) {
  console.log(JSON.stringify(fields));
}

export async function completeWithFallback({ tenantId, scenario, messages }) {
  const requestId = crypto.randomUUID();

  for (const [fallbackIndex, route] of routes[scenario].entries()) {
    for (let attempt = 1; attempt <= 2; attempt += 1) {
      const started = Date.now();
      try {
        const response = await client.chat.completions.create({
          model: route.model,
          messages,
          timeout: route.timeoutMs,
        });
        logEvent({
          event: "llm_call_ok",
          request_id: requestId,
          tenant_id: tenantId,
          scenario,
          model: route.model,
          cost_group: route.group,
          fallback_index: fallbackIndex,
          attempt,
          latency_ms: Date.now() - started,
        });
        return response.choices[0].message.content;
      } catch (error) {
        const status = error.status || error.code;
        logEvent({
          event: "llm_call_error",
          request_id: requestId,
          tenant_id: tenantId,
          scenario,
          model: route.model,
          cost_group: route.group,
          fallback_index: fallbackIndex,
          attempt,
          status,
        });
        if (status === 429 || status === "ETIMEDOUT" || status >= 500) {
          await sleep(Math.min(1000 * 2 ** attempt, 6000));
          continue;
        }
        throw error;
      }
    }
  }

  throw new Error(`all LLM routes failed: request_id=${requestId}`);
}

if (import.meta.url === `file://${process.argv[1]}`) {
  const content = await completeWithFallback({
    tenantId: "tenant_demo",
    scenario: "internal_tool",
    messages: [{ role: "user", content: "Draft a concise customer support macro." }],
  });
  console.log(content);
}
